### ISSUE TITLE
allow module descriptor to override dependency versions

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-spec#1134]
We've talked about letting dependency version conflicts be resolved at the assembly level. But in fact people building reusable libs don't package their lib as an assembly so we need some way to resolve version conflicts at the module level in the module descriptor. What should this look like?

**Comment by luolong** [comment]
It seems that this discussion has moved to  #1185

**Comment by gavinking** [comment]
> `override import` is not really a dependency. What if I change my direct dependencies such that none of them require xoo.baz, should module loader still download it? 

Does it actually _hurt_ if it does?

> My code directly depends only on the SLF4J API and it's needed for compiling the module. But, this API needs some kind of implementation library at runtime. I do not really care at this point which though.

Isn't this _exactly_ the usecase which assemblies seek to address?

**Comment by gavinking** [comment]
@jvasileff scanning your post, I ask myself if it isn't the case that some of those problem would be solved if we had a way to just throw some modules on a single shared classloader that is accessible to all other modules. Sorta like an emulated "flat classpath".

**Comment by jvasileff** [comment]
@gavinking I assume you're referring to my Nov 18th post? I believe that post is full of opinions that would naturally lead more towards a one-version-per-module result and therefore a minimal number of classloaders, but it wasn't a call for a flat classpath per se.

I see the classloader hierarchy as a logical outcome of version resolution. If for some reason we *wanted* a flat classpath (which would be a separate topic), then I think we'd have to create versioning and import rules that would lead to that result.

If you need to load two classes with the same name, you need two classloaders. If you don't, you don't. (Sure, JBoss modules may throw in a few extra class loaders for isolation, but that's beside the point.)

Maybe you're thinking of a work-around? But so far, I've constrained my thoughts/comments to non-compromised scenarios. No module gets to break the rules and do things like "just throw in some extra classes". So even modules that consider themselves to be applications should be importable by other modules.

**Comment by LasseBlaauwbroek** [comment]
In my view, the only correct way to manage non-exported dependencies is to _always_ load them with separate classloaders. Even if there are multiple non-exported dependencies on the same module and version. This way, you make sure that modules can not interfere with eachother using variable toplevels as @jvasileff mentions in the [mailinglist](https://groups.google.com/forum/#!topic/ceylon-users/3xMraYFeHG0). If you use a logger in your module and want it to be configured by another module, you should make the logger shared.

The only downside to this policy I can see is that it can get very bloated with a lot of the same non-shared modules loaded. A solution to this can be that a module author can indicate that it's module does not have any global state and can be loaded only once. I guess this cannot be mechanically checked :-(

**Comment by sgalles** [comment]
> If for some reason we wanted a flat classpath (which would be a separate topic)

* Yes @jvasileff indeed, a flat classpath feature is a separate topic, but I think that it would be immediately useful.
At least the proposal of @gavinking is straightforward and would quickly allow a "flat classpath for dummies" that IMHO would make interop POC much more pleasant. And actually I believe that even with an advanced classpath scheme, I would still use the "flat classpath" to perform basic interop tests (and then I'll try to switch to the advanced module scheme if necessary).

**Comment by jvasileff** [comment]
@sgalles My intention was to stress the point that the classloader arrangement is not an independent choice, but is rather dictated by the mix of modules that you need to load. So if for some reason (runtime environment, or whatever) you wanted to have only a single classloader, you would need *stricter* rules. That is, one version per module, throughout.

But I think I understand your point too - some way to formalize or automate the fat-jar approach? I think all of these issues can be *very* complicated and easily intermixed. Perhaps a separate issue or thread would make sense to discuss shorter term workarounds?

**Comment by sgalles** [comment]
> But I think I understand your point too - some way to formalize or automate the fat-jar approach

Yes indeed @jvasileff . That's the "flat classpath" proposed by @gavinking 
 And you're perfectly right, that's probably an other issue (I prefer not  to create the issue though, @gavinking will see what he wants to do with all this)

**Comment by luolong** [comment]
> > override import is not really a dependency. What if I change my direct dependencies such that none of them require xoo.baz, should module loader still download it?
> Does it actually hurt if it does?

Well, there's bandwidth and memory consumption concerns...
